### DISCUSSION TITLE
Proposal: Allow Expression trees in Attribute constructor

### DISCUSSION THREAD

**Discussion by burtonrodman** [dotnet/csharplang#6732] [Language Ideas] [answered]
Problem(s):
- I would like to provide a binding expression in an attribute of my model.
- I would like to provide an expression that my Source Generator will use to extract a value in a generated constructor (think extracting Value from IOptions in AspNetCore).
- I would like to declare an Expression<Func<T1, T2>> and provide T1, but have the compiler infer T2 from the body of the expression.

Proposed Changes to C#:
- allow expression trees in Attribute constructors
- allow "partial inference" of generic type parameters
 
```
[GenerateServiceConstructor]
public partial class Test
{
    // Attribute constructor parameter 'initAction' has type 'Expression<Func<IOptions<TestOptions>, TestOptions>>', which is not a valid attribute parameter type csharp(CS0181)
    [InjectAs<IOptions<TestOptions>, TestOptions>(o => o.Value)]
    public required TestOptions TestOptions { private get; init; }
}

[AttributeUsage(AttributeTargets.Property | AttributeTargets.Field)]
public class InjectAsAttribute<TInject, TResult> : Attribute
    where TInject : class
    where TResult : class
{
    public InjectAsAttribute(
        Expression<Func<TInject, TResult>> initAction
    )
    {
    }
} 
```

```
// Using the generic type 'InjectAsAttribute<TInject, TResult>' requires 2 type arguments csharp(CS0305)
var foo = new InjectAsAttribute<IOptions<TestOptions>>(o => o.Value);
```

**Comment by HaloFour** [comment, marked as answer]
> allow expression trees in Attribute constructors

See: #343

> allow "partial inference" of generic type parameters

See: #1018, #997
